waku/v2/waku_archive/archive.nim

```diff
@@ -186,25 +231,140 @@ proc findMessages*(w: WakuArchive, query: ArchiveQuery): Future[ArchiveResult] {
 
 # Retention policy
 
-proc executeMessageRetentionPolicy*(w: WakuArchive) {.async.} =
+proc executeMessageRetentionPolicy*(w: WakuArchive):
+                                    Future[Result[void, string]] {.async.} =
   if w.retentionPolicy.isNil():
-    return
+    return err("retentionPolicy is Nil in executeMessageRetentionPolicy")
 
   if w.driver.isNil():
-    return
+    return err("driver is Nil in executeMessageRetentionPolicy")
 
   let retPolicyRes = await w.retentionPolicy.execute(w.driver)
   if retPolicyRes.isErr():
       waku_archive_errors.inc(labelValues = [retPolicyFailure])
-      error "failed execution of retention policy", error=retPolicyRes.error
+      return err("failed execution of retention policy: " & retPolicyRes.error)
 
-proc reportStoredMessagesMetric*(w: WakuArchive) {.async.} =
+  return ok()
+
+proc reportStoredMessagesMetric*(w: WakuArchive):
+                                 Future[Result[void, string]] {.async.} =
   if w.driver.isNil():
-    return
+    return err("driver is Nil in reportStoredMessagesMetric")
 
   let resCount = await w.driver.getMessagesCount()
   if resCount.isErr():
-    error "failed to get messages count", error=resCount.error
-    return
+    return err("failed to get messages count: " & resCount.error)
 
   waku_archive_messages.set(resCount.value, labelValues = ["stored"])
+
+  return ok()
+
+proc startMessageRetentionPolicyPeriodicTask*(w: WakuArchive,
+                                              interval: timer.Duration) =
+  # Start the periodic message retention policy task
+  # https://github.com/nim-lang/Nim/issues/17369
+
+  var executeRetentionPolicy: proc(udata: pointer) {.gcsafe, raises: [Defect].}
+  executeRetentionPolicy = proc(udata: pointer) {.gcsafe.} =
+
+    try:
+      let retPolRes = waitFor w.executeMessageRetentionPolicy()
+      if retPolRes.isErr():
+        waku_archive_errors.inc(labelValues = [retPolicyFailure])
+        error "error in periodic retention policy", error = retPolRes.error
+    except CatchableError:
+      waku_archive_errors.inc(labelValues = [retPolicyFailure])
+      error "exception in periodic retention policy",
+            error = getCurrentExceptionMsg()
+
+    discard setTimer(Moment.fromNow(interval), executeRetentionPolicy)
+
+  discard setTimer(Moment.fromNow(interval), executeRetentionPolicy)
+
+proc setupWakuArchiveDriver(url: string, vacuum: bool, migrate: bool):
```

Comment:
This is application logic: decisions about how the configured url should be interpreted and which specific backend storage systems should be set up for the application lies with the application. When adding new drivers (e.g. someone now supports Codex distributed storage) the point of a driver interface is exactly that the backend operation would not have to be changed or even be aware of this fact. In this case, the application would have to request changes in the setup instructions which couples the driver interface with application-defined underlying implementations.